### PR TITLE
Refactor.lint

### PR DESCRIPTION
**Pull request by wxtim**
I did a review of my older code as a training exercise, and started refactoring.

- The big win - remove duplication by merging `get_reference_text` and `get_reference_rst`.
- Turned on the Hyperlinks in the documentation. Test this by building Cylc Doc.
- Spun as much as possible out of the main function.
- Added typing to more functions.

There's plenty more refactoring I'd like to do, but I really shouldn't be prioritizing this.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes - I have deliberately created a chain of changes in individual commits - feel free to squash-merge.
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are minimally changed.
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by wxtim** [comment]
@oliver-sanders  - Draft because failing tests, or because low priority, or some other reason?

File: cylc/flow/scripts/lint.py [hunk 1]
```diff
@@ -1025,88 +1059,106 @@ def get_cylc_files(
...
-            url = get_url(meta)
-            summary = meta.get("rst", meta['short'])
+            template = issue_heading_template
+            url = meta.get('url', '')
+            if url and url.startswith('http'):
+                url = meta['url']
+            elif url:
+                url = (
+                    "https://cylc.github.io/cylc-doc/stable/html/7-to-8/"
+                    + url)
```
[the remark this answers is not in the document]
Comment (wxtim) [reply cylc/flow/scripts/lint.py]: Well spotted.

File: cylc/flow/scripts/lint.py [hunk 2]
```diff
@@ -1271,6 +1282,4 @@ def main(parser: COP, options: 'Values', target=None) -> None:
 
 # NOTE: use += so that this works with __import__
 # (docstring needed for `cylc help all` output)
-__doc__ += TOMLDOC.format(
-    '\n\n.. code-block:: toml\n', str(LINT_SECTIONS)) + get_reference_rst(
-    parse_checks(['728', 'style'], reference=True))
+__doc__ += get_reference('all', 'rst')
```
Comment (MetRonnie) [inline cylc/flow/scripts/lint.py:1285]: Is this supposed to be
```python
__doc__ += TOMLDOC.format(
    '\n\n.. code-block:: toml\n', str(LINT_SECTIONS)
) + get_reference('all', 'rst')
```
?

Comment (MetRonnie) [reply cylc/flow/scripts/lint.py:1285]: Actually I think not, this was what is currently causing cylc-doc build failures but your changes fix it

Comment (MetRonnie) [reply cylc/flow/scripts/lint.py:1285]: (Will need a fix backported to 8.2.x)

**Review by markgrahamdawson — APPROVED** [review]

**Pull request merged**